OrbitService/ProducerSideServiceImpl.cpp

```diff
@@ -0,0 +1,324 @@
+  receive_events_thread.join();
+
+  // When receive_events_thread exits because stream->Read(&request) fails,
+  // it means that the producer has disconnected: ask send_commands_thread to exit, too.
+  exit_send_commands_thread = true;
+  send_commands_thread.join();
+
+  {
+    absl::MutexLock lock{&server_contexts_mutex_};
+    server_contexts_.erase(context);
+  }
+
+  LOG("Finished handling ReceiveCommandsAndSendEvents for a CaptureEventProducer");
+  return grpc::Status::OK;
+}
+
+void ProducerSideServiceImpl::SendCommandsThread(
+    grpc::ServerContext* context,
+    grpc::ServerReaderWriter<orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse,
+                             orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest>* stream,
+    bool* all_events_sent_received, std::atomic<bool>* exit_send_commands_thread) {
+  // As a result of initializing prev_capture_status to kCaptureFinished,
+  // an initial StartCaptureCommand is sent
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStarted,
+  // and an initial StopCaptureCommand is sent (with little effect)
+  // if service_state_.capture_status is actually CaptureStatus::kCaptureStopping.
+  CaptureStatus prev_capture_status = CaptureStatus::kCaptureFinished;
+
+  // This loop keeps track of changes to service_state_.capture_status using
+  // conditional critical sections on service_state_mutex_ and updating prev_capture_status,
+  // and sends StartCaptureCommands and StopCaptureCommands accordingly.
+  // It exits when *exit_send_commands_thread is true, when service_state_.exit_requested is true,
+  // or when Write fails (because the producer disconnected or because the context was cancelled).
+  while (true) {
+    // This is set when ReceiveEventsThread has exited. At that point this thread should also exit.
+    if (*exit_send_commands_thread) {
+      return;
+    }
+
+    service_state_mutex_.Lock();
+    if (service_state_.exit_requested) {
+      service_state_mutex_.Unlock();
+      return;
+    }
+
+    if (service_state_.capture_status != prev_capture_status) {
+      switch (service_state_.capture_status) {
+        case CaptureStatus::kCaptureStarted: {
+          prev_capture_status = service_state_.capture_status;
+          ++service_state_.producers_remaining;
+          *all_events_sent_received = false;
+          service_state_mutex_.Unlock();
+
+          orbit_grpc_protos::ReceiveCommandsAndSendEventsResponse command;
+          command.mutable_start_capture_command();
+          if (!stream->Write(command)) {
+            ERROR("Sending StartCaptureCommand to CaptureEventProducer");
+            LOG("Terminating call to ReceiveCommandsAndSendEvents as Write failed");
+            // Cause Read in ReceiveEventsThread to also fail if for some reason it hasn't already.
+            context->TryCancel();
+            return;
+          }
+          LOG("Sent StartCaptureCommand to CaptureEventProducer");
+        } break;
```

Comment:
nit: move break inside {}?

Reply:
The break outside is something we do elsewhere and that I like. The idea is to separate the actual code from the `break` required by the old `switch` construct, which has nothing to do with the content of the braces.